cpp/src/arrow/io/compressed.cc

```diff
@@ -397,10 +415,11 @@ class CompressedInputStream::Impl {
   std::shared_ptr<InputStream> raw_;
   bool is_open_;
   std::shared_ptr<Decompressor> decompressor_;
-  std::shared_ptr<Buffer> compressed_;
+  std::shared_ptr<ResizableBuffer> compressed_;
   // Position in compressed buffer
   int64_t compressed_pos_;
   std::shared_ptr<ResizableBuffer> decompressed_;
+  std::shared_ptr<ResizableBuffer> decompressed_impl_;
```

Comment:
Why are there two buffers suddently? What is the difference between the two?

Reply:
It's not neccessary to maintain two buffer, I use it here for simplify the changes(previous impl set `decompressed_ = nullptr` to means decompressed is consumed). Seems this is not a good idea.